candies/cli/args/typed.py

```diff
@@ -0,0 +1,47 @@
+from typing import Type, Any
+
+
+class Typed:
+    """A typed command line argument.
+
+    A typed command line argument is used for specifying the type
+    to convert the value to. For example, consider the following code:
+
+        @cli
+        def double(number: Opt[int]('A number to double.')):
+            print(number * 2)
+
+    When this CLI is called as
+
+        $ example.py double --number 5
+
+    the value '5' of the argument '--number' is converted to `int`
+    and passed to the function.
+
+    Note:
+        The default type of arguments is `str`. Thus, if an argument
+        is annotated as `Opt('A number to double.')`, a string value
+        will be passed to the function.
+
+    Attributes:
+        kind: A kind of an argument
+            (for example, `Arg`, `Opt` or `Flag`).
+        type: A type of an argument
+            (for example, `int`, `List[double]`, etc.).
+    """
+
+    def __init__(self, kind: Type, type: Type):
+        self.kind = kind
+        self.type = type
+
+    def __call__(self, *args: Any, **kwargs: Any) -> 'Typed':
+        # Instantiate the parameter being wrapped. For example,
+        # `Arg[int]` will return `Typed(Arg, int)`, and
+        # `Typed(Arg, int)('Description.')` will instantiate
+        # `self.kind = Arg('Description.')`.
+        if isinstance(self.kind, type):
+            self.kind = self.kind(*args, **kwargs)
+        else:
+            raise ValueError(f'Cannot instantiate {type(self.kind)} twice.')
```

Comment:
I'm wondering about the following example:
```py
def main(x: Arg[int]('Description')(), y: Arg): ...
```

Will the message contain the name of the argument `x`? Or we need to implement `__str__` for each concrete argument kind `Arg` ,`Opt` and `Flag`?

Reply:
No, unfortunately, at the moment of instantiation we don't know the name of the argument, and thus, we cannot print it.

Reply:
Well, at least can we show the `__repr__` of an object? For example, `Cannot instantiate {type(self.kind)} twice: it's already instantiated as {repr(self.kind)}`.

Reply:
Sure!